### DISCUSSION TITLE
"Proper" way to build Ghidra while iterating small changes?

### DISCUSSION THREAD

**Discussion by hippietrail** [NationalSecurityAgency/ghidra#7038] [Q&A]
I'm an old programmer who barely ever touched Java, now doing some Ghidra programming for fun.

I know learning how to use Java "properly" is a gigantic rabbithole so I just want to ask if I'm building it the right way or if there's something much smaller and faster I should be using?

Last week I downloaded the source and built it for the first time while doing other things and it worked fine. Then I put it away until today. Today I changed the name of one function to see how many errors would show up and built it the same way as last time, which I found on the Internet somewhere:
```
export PATH=$PATH:/Users/hippietrail/Downloads/gradle-8.10.2/bin
gradle buildGhidra
```
I was expecting it to be much fast since all the setup was done and a first build was done. It's on an external hard drive of my M1 MacBook Air. It took just over 20 minutes!

Can I do anything to make this fast enough to iterate experimental code changes?

**Comment by ryanmkurtz** [comment]
If you want to do fast iterative development in the java code, i recommend you [follow our instructions for setting up Eclipse](https://github.com/NationalSecurityAgency/ghidra#advanced-development).  That's how the team develops.

**Reply by hippietrail** [reply]
Thanks! I had looked for official build instructions a couple of times and failed to find them. But I finally did and came back to delete my question here...

**Reply by hippietrail** [reply]
I've followed those instructions and have a Ghidra build now on an external macos-formatted hard drive. I can set breakpoints and step through. But when I modify the source at `Ghidra/Features/Base/src/main/java/ghidra/plugins/fsbrowser/FSBComponentProvider.java` the running Ghidra is now out of sync with the code in the source. New function calls and variables I introduce are ignored by the code running or single-stepping. If I intentionally add syntax errors, Eclipse marks them as errors but Ghidra still happily runs the original code out of sync with what Eclipse is showing.

Quitting and restarting Eclipse doesn't make a difference. Doing `gradle prepdev eclipse buildNatives` again in the console takes some time but makes no difference.

Selecting everything in the Project Explorer and doing "refresh" then starting again made no difference.

**Reply by Wall-AF** [reply]
Are you changing the code within Eclipse?  If so, is it rebuilding automatically?  If not then refresh and rebuild Eclipse.

**Reply by ryanmkurtz** [reply]
When you say you have a Ghidra build, what do you mean? When developing out of eclipse, there is just the git repo...there shouldn't be another "build" somewhere.

**Reply by ryanmkurtz** [reply]
How are you launching Ghidra in this scenario?

**Reply by hippietrail** [reply]
> Are you changing the code within Eclipse? If so, is it rebuilding automatically? If not then refresh and rebuild Eclipse.

Yes. When I was working on extensions I was switching between VS Code and Eclipse but now trying to work on the main repo I'm just trying to do it in Eclipse so far.

**Reply by hippietrail** [reply]
> When you say you have a Ghidra build, what do you mean? When developing out of eclipse, there is just the git repo...there shouldn't be another "build" somewhere.

I mean the app built successfully with gradle between just cloning the repo and trying to add some code changes.
I do also have the Ghidra I downloaded but when I start it up it gives a version one point newer than the latest release so I know it the one I built running and not the one I downloaded.

**Reply by hippietrail** [reply]
> How are you launching Ghidra in this scenario?

From the debug icon and choosing "Ghidra".
